osquery/tables/system/darwin/user_groups.mm

```diff
@@ -0,0 +1,107 @@
+/*
+ *  Copyright (c) 2014, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#include <vector>
+#include <string>
+
+#include <grp.h>
+#include <pwd.h>
+
+#import <OpenDirectory/OpenDirectory.h>
+
+#include <osquery/core.h>
+#include <osquery/filesystem.h>
+#include <osquery/logger.h>
+#include <osquery/tables.h>
+
+#include "osquery/tables/system/user_groups.h"
+
+namespace osquery {
+namespace tables {
+
+QueryData genUserGroups(QueryContext &context) {
+  @autoreleasepool {
+    QueryData results;
+
+    ODSession *session = [ODSession defaultSession];
+    NSError *err;
+    ODNode *root =
+        [ODNode nodeWithSession:session name:@"/Local/Default" error:&err];
+    if (err) {
+      LOG(ERROR) << "Error with OD node: "
+                 << std::string([[err localizedDescription] UTF8String]);
+      return results;
+    }
+    ODQuery *q = [ODQuery queryWithNode:root
+                         forRecordTypes:kODRecordTypeUsers
+                              attribute:nil
+                              matchType:0
+                            queryValues:nil
+                       returnAttributes:nil
+                         maximumResults:0
+                                  error:&err];
+    if (err) {
+      LOG(ERROR) << "Error with OD query: "
+                 << std::string([[err localizedDescription] UTF8String]);
+      return results;
+    }
+
+    NSArray *od_results = [q resultsAllowingPartial:NO error:&err];
+    if (err) {
+      LOG(ERROR) << "Error with OD results: "
+                 << std::string([[err localizedDescription] UTF8String]);
+      return results;
+    }
+
+    int groups_buf[EXPECTED_GROUPS_MAX];
+    int *groups;
+    int ngroups;
+    for (ODRecord *re in od_results) {
+      std::string username = std::string([[re recordName] UTF8String]);
+      struct passwd *pwd = getpwnam(username.c_str());
+      if (pwd != nullptr) {
+        ngroups = EXPECTED_GROUPS_MAX;
+        groups = groups_buf;
+  
+        // GLIBC version before 2.3.3 may have a buffer overrun:
+        // http://man7.org/linux/man-pages/man3/getgrouplist.3.html
+        if (getgrouplist(pwd->pw_name, pwd->pw_gid, groups, &ngroups) < 0) {
+          // EXPECTED_GROUPS_MAX was probably not large enough.
+          // Try a larger size buffer.
+          // Darwin appears to not resize ngroups correctly.  We can hope
+          // we had enough space to start with.
+          LOG(ERROR) << "Resizing to " << ngroups;
```

Comment:
Mind changing all the `LOG(ERROR)` calls to `TLOG`? (tlog for table log... haha) Right now the `TLOG` macro expands to the `VLOG(1)` macro. In the future we might build in more controls around logs from table-implementation code.